ion-schema/src/ordered_elements_nfa.rs

```diff
@@ -0,0 +1,443 @@
+                    min_visits,
+                    max_visits,
+                    description,
+                };
+
+                states.push(State::Intermediate(i + 1, state))
+            });
+
+        // This will become the ID of the "Final" state, but it's convenient to wait to add the
+        // "Final" state until we've determined the edges of the graph.
+        let max_id = states.len();
+
+        // Construct an adjacency list to represent the edges of the graph.
+        let mut edges = vec![];
+        for (i, s) in states.iter().enumerate() {
+            // Loop back to self, if max is > 1
+            let min_transition = if s.can_reenter(1) { i } else { i + 1 };
+
+            // Add transitions forward up to (including) the first type with a min occurs that is greater than 0
+            let mut j = i + 1;
+            while j < max_id {
+                let state_j = states.get(j).unwrap();
+                if !state_j.can_exit(0) {
+                    break;
+                }
+                j += 1;
+            }
+            edges.push(min_transition..=j)
+        }
+
+        // We make `Final` have fake edges to `Initial`. `Initial` denies all entry and `Final`
+        // doesn't allow exiting, so it's not actually traversable, and besides, you can't enter
+        // `Final` unless you're at the end-of-sequence event, so there's no more input to handle
+        // that would make it even try any further transitions.
+        states.push(State::Final(max_id));
+        edges.push(0..=0);
+
+        // Terminal state is the Final state with a visit count of 1.
+        let terminal_state: StateVisitCount = (max_id, 1usize);
+
+        OrderedElementsNfa {
+            states,
+            edges,
+            terminal_state,
+        }
+    }
+
+    /// Tests an input sequence of [Element]
+    pub fn matches<'a, I: Iterator<Item = &'a Element>>(
+        &self,
+        mut iter: I,
+        type_store: &'a TypeStore,
+        ion_path: &mut IonPath,
+    ) -> ValidationResult {
+        let mut current_state_set: HashSet<StateVisitCount> = HashSet::new();
+        let mut new_states: HashSet<StateVisitCount> = HashSet::new();
+
+        let mut input_index = 0;
+        current_state_set.insert((0usize, 1usize));
+
+        // Essentially, for-each input, but we want to capture the `Option::None` at the end of the iterator.
+        loop {
+            let element: Event = iter.next();
+            let mut invalid_transitions: HashSet<TraversalError> = HashSet::new();
+
+            ion_path.push(IonPathElement::Index(input_index));
+
+            // For each state in the set of current states...
+            for &(from_state_id, num_visits) in &current_state_set {
+                let from_state = self.states.get(from_state_id).unwrap();
+
+                // Need to clone the range because strangely &RangeInclusive doesn't
+                // implement Copy or IntoIterator.
```

Comment:
Happily, this weirdness is [getting addressed in edition 2024](https://github.com/rust-lang/rfcs/blob/master/text/3550-new-range.md).